馬別統計データの表示

mterada1228 opened the pull request.

mterada1228 committed:
デザイン案

mterada1228 committed:
Merge branch 'develop' into features/show_stats_of_horse

mterada1228 committed:
馬別統計情報（馬場別）出力処理のサーバサイドを開発

ershn commented on app/facades/horse_show.rb:
```diff
@@ -0,0 +1,19 @@
+      ]
+    end.to_h
+  end
+end
```
ここの処理は馬のレース結果の統計をまとめたオブジェクトと、そのオブジェクトを生成する処理に分けるのがいいかと思いました。

クラス名・インターフェイスとしては
- `HorseEvaluation`
- `EvaluateHorse.new(horse).call`

がいいかと思います（馬を評価するところから）。

コントローラーではサービスを使うことで`HorseEvaluation`オブジェクトを生成して、ビューに直接渡して問題ないかと。

mterada1228 replied:
馬別統計データを計算して、`HorseEvaluation` のインスタンスを作成する `EvaluateHorse` クラスを追加しました。  
`HorseController` クラスでは、`EvaluateHorse.new(@horse).call` をして、`HorseEvaluation` のインスタンスを取得します。  
View では `HorseEvaluation` のプロパティを参照して、馬別統計データの表示を実現するように修正しました。

ershn replied:
OK

ershn commented on app/facades/horse_show.rb:
```diff
@@ -0,0 +1,19 @@
+          third: races.third_place_races.count,
+          unplaced: races.unplaced_races.count
+        }
+      ]
```
上の処理で SQL クエリが多く実行されてしまうと思いますが、それだと、パフォーマンスが問題になる可能性があります。

必要なデータすべてを取得してから、rails 側でデータの加工をした方がいいかもしれません。

mterada1228 replied:
以下の処理に変更することで、SQLクエリは 16回 => 4回に削減できました。
- `EvaluateHorse#races_by_place` で初めに、その馬の1着、2着、3着、着外の `Race` を `@races_by_place` として取得する
- `EvaluateHorse#results_by_course_condition` => `create_results_by_value_hash` では、`@races_by_place` の中から、`course_condition` が　`firm, good, yielding, soft` となっているレースを`Enumerable#select` で抽出して、`course_condition` 毎の 1着, 2着, 3着, 着外の回数の `Hash` を作成する

ershn replied:
OK

ershn commented on app/models/race.rb:
```diff
@@ -19,6 +19,12 @@ class Race < ApplicationRecord
+  scope :first_place_races, -> { joins(:horse_races).merge(HorseRace.first_place_races) }
+  scope :second_place_races, -> { joins(:horse_races).merge(HorseRace.second_place_races) }
+  scope :third_place_races, -> { joins(:horse_races).merge(HorseRace.third_place_races) }
+  scope :unplaced_races, -> { joins(:horse_races).merge(HorseRace.unplaced_races) }
```
わかりやすくなるように、名前を工夫した方がいいかと思いました。

mterada1228 replied:
`Race` なのか `HorseRace` なのか、どのモデルを取得しようとしているのかがわかりづらい。

mterada1228 replied:
取得すべきモデルは `HorseRace` が主でなので、`hoge_place_horse_races` に修正しました。

ershn replied:
OK

mterada1228 committed:
refs features/show_stats_of_horse 指摘対応
  - RaceResult モデルは常に Race が必要となるため廃止して、Raceモデルに統合
  - Raceモデルでは、レース結果（旧RaceResul）が提供されているかどうかを`has_result` column で制御する

mterada1228 reviewed (commented):
`RaceResult` を `Race` に統合したことに関して、ただ単に `RaceResult` から `Race` モデルを参照するように修正した箇所と、データマイグレーションに関連するファイルは本MRと本質的には関係がないため一旦 レビューステータスを`Viewed` にしておきました。必要に応じてご確認頂ければと思います。

ershn commented on app/services/evaluate_horse.rb:
```diff
@@ -0,0 +1,47 @@
+                     second_place_races: @horse.races.second_place_horse_races,
+                     third_place_races: @horse.races.third_place_horse_races,
+                     unplaced_races: @horse.races.unplaced_horse_races })
+  end
```
ここで SQL クエリは4つ発行されますが、全部同じテーブルで条件はほぼ一緒なので、一つの SQL クエリで全部を取得してから、Rails で処理をしたほうがパフォーマンス的にいいと思います（group_byなど）。

mterada1228 replied:
SQL クエリの発行は1回のみ、https://github.com/mterada1228/keiba_expectation_app/blob/53a48c68850528d0a40ebdc09fea16b639ce00ad/app/services/horse_race_stats.rb#L3 で実行するように修正しました。`order_of_arrival` での絞り込み、`course_condition` での絞り込みは、ruby のメソッドで行うようにしております。

ershn replied:
OK

ershn commented on app/services/evaluate_horse.rb:
```diff
@@ -0,0 +1,47 @@
+                    race.send(column) == value
+                  end.count }
+    )
+  end
```
ここで事前に準備された`@races_by_place`を使ってますが、メソッドのインターフェイスでそれがわからなくて、`@races_by_place`がない状態で読んでしまうと、バグってしまうので、`@races_by_place`は引数から受け取った方が保守性的にいいです（データのフローがわかりやすくなるので）。

mterada1228 replied:
この関数自体は削除してしまいましたが、相当の処理は以下に移動しました。

https://github.com/mterada1228/keiba_expectation_app/blob/53a48c68850528d0a40ebdc09fea16b639ce00ad/app/services/horse_race_stats.rb#L10

ここではインスタンス変数を直接呼び出しせず、引数として受け取るように変更しています。

ershn replied:
OK

ershn commented on app/services/evaluate_horse.rb:
```diff
@@ -0,0 +1,47 @@
+  def create_results_by_value_hash(column, value)
+    OpenStruct.new(
+      { first_place: @races_by_place.first_place_races.select do |race|
+                       race.send(column) == value
```
Rails のカラムは `race[column]` の形でもアクセスできます。

mterada1228 replied:
こちらの処理は `group_by` で行うように変更しましたので、該当のメソッドは削除されました。

https://github.com/mterada1228/keiba_expectation_app/blob/53a48c68850528d0a40ebdc09fea16b639ce00ad/app/services/horse_race_stats_value.rb#L5

ershn replied:
OK

ershn commented on app/services/evaluate_horse.rb:
```diff
@@ -0,0 +1,47 @@
+    OpenStruct.new(
+      { first_place: @races_by_place.first_place_races.select do |race|
+                       race.send(column) == value
+                     end.count,
```
`count`にブロックを直接渡すことも可能です。

mterada1228 replied:
該当の処理は`group_by`の結果を`count`するよう変更したので、該当のメソッドは削除しました。

https://github.com/mterada1228/keiba_expectation_app/blob/53a48c68850528d0a40ebdc09fea16b639ce00ad/app/services/horse_race_stats_value.rb#L6-L8

ershn replied:
OK

ershn commented on app/services/evaluate_horse.rb:
```diff
@@ -0,0 +1,47 @@
+  end
+
+  def create_results_by_value_hash(column, value)
+    OpenStruct.new(
```
OpenStruct は避けて`HorseEvaluation`と同じく名前を与えたクラスにした方がいいと思います。

mterada1228 replied:
以下の統計情報のオブジェクトをクラスとして設計しました。

1. 統計情報オブジェクト `HorseRaceStats`
2. 各統計上の値のオブジェクト `HorseRaceStatsValue`

例) `Race.course_conditions` 毎の統計情報オブジェクト
``` ruby
<HorseEvaluation:0x00007fe05d5c65f8
 @results_by_course_condition=
  #<HorseRaceStats:0x00007fe05d5c4550
   @firm=
    #<HorseRaceStatsValue:0x00007fe05c5f1768 @first_place=4, @second_place=4, @third_place=3, @unplaced=13>,
   @good=#<HorseRaceStatsValue:0x00007fe05f9afc08 @first_place=1, @second_place=1, @third_place=0, @unplaced=2>,
   @soft=#<HorseRaceStatsValue:0x00007fe060126ec8 @first_place=0, @second_place=0, @third_place=0, @unplaced=2>,
   @yielding=
    #<HorseRaceStatsValue:0x00007fe05f914cf8 @first_place=0, @second_place=0, @third_place=1, @unplaced=1>>>

ershn replied:
OK

ershn commented on app/services/evaluate_horse.rb:
```diff
@@ -0,0 +1,47 @@
+                     end.count,
+        unplaced: @races_by_place.unplaced_races.select do |race|
+                    race.send(column) == value
+                  end.count }
```
以上で select は全部いっしょなので、select を実行した上で順位でフィルターした方がいいと思います。

mterada1228 replied:
順位のフィルターは、`HorseRace` の `order_of_arrival` で `group_by` することで取得するように変更しました。

https://github.com/mterada1228/keiba_expectation_app/blob/53a48c68850528d0a40ebdc09fea16b639ce00ad/app/services/horse_race_stats_value.rb#L5

ershn replied:
OK

mterada1228 committed:
refs features/show_stats_of_horse 指摘対応

mterada1228 commented:
@ershn 修正を行いましたので、再レビューをお願いいたします。

ershn commented on app/services/horse_race_stats.rb:
```diff
@@ -0,0 +1,34 @@
+      end
+    end
+  end
+end
```
汎用的ですが、逆に読みにくくて保守するのも難しいと思います。

ここは属性ごとに個別なクラスを作るのは問題ないと思います。

`define_method`や`instance_variable_set`、`class_eval`などは他にやり方が本当にない時だけ使うべきです。

mterada1228 replied:
個別のクラス `ResultsByCourseCondition` を作成して、メタプログラミング的な記法を避けるようにしました。  
ただ、`instance_variable_set` の箇所は他に良い書き方が思いつかないのですが、代わりに適した記法を相談させていただけないでしょうか...？

ershn replied:
OK

ershn commented on app/services/horse_race_stats.rb:
```diff
@@ -0,0 +1,34 @@
+class HorseRaceStats
+  def initialize(horse:, column:)
+    horse_races = horse.horse_races.includes(:race)
```
`includes`の呼び出し順番が合ってますか？

mterada1228 replied:
リファクタリングをして街頭のメソッドは以下に移動しました。
https://github.com/mterada1228/keiba_expectation_app/blob/b361b6da0c675a2f87cb53a0f7094d3b05e60a70/app/controllers/horses_controller.rb#L9-L11

Horse => Race, HorseRace への has_many 関連と、HorseRace => Race への belongs_to の関連の2種類があるので、`horse_races.includes(:race)` は必要なものになります。

ershn replied:
OK

ershn replied:
OK

ershn commented on app/services/horse_race_stats_value.rb:
```diff
@@ -0,0 +1,19 @@
+class HorseRaceStatsValue
+  attr_accessor :first_place, :second_place, :third_place, :unplaced
```
`attr_reader`の方が適切だと思います。

mterada1228 replied:
こちら `attr_reader` に修正いたしました。

ershn replied:
OK

ershn commented on app/services/horse_race_stats_value.rb:
```diff
@@ -0,0 +1,19 @@
+class HorseRaceStatsValue
```
クラス名は汎用的すぎて、何を指すのかがよくわからないので、もっと具体的な`HorsePodiums`などの方が適切だと思います。

mterada1228 replied:
`HorsePodiums` に修正いたしました。

ershn replied:
OK

mterada1228 committed:
refs features/show_stats_of_horse 指摘対応

mterada1228 committed:
refs features/show_stats_of_horse add env_files to gitignore

ershn commented on app/services/generate_rpci_ave1f_scatter_plots_service.rb:
```diff
@@ -27,7 +27,7 @@ def generate_scatter_plot(horse_race)
   def scatter_plot(horse_race, color)
     { name: horse_race.race.name,
-      data: [[horse_race.race.race_result.RPCI, horse_race.race.race_result.ave_1F]],
+      data: [[horse_race.race.RPCI, horse_race.race.ave_1F]],
```
ここで`race`の取得で不要な SQL が実行されないように、リレーションの設定で`inverse_of`を生かす必要があるかもしれません。

mterada1228 replied:
すみません、嘘つきました...  
Rails 4.1 以降では `inverse_of` を明示的に示す必要はないようです。

実際に試しましたが、複数回の SQL は実行されませんでした。

ershn replied:
OK

ershn commented on app/services/horse_stats/evaluate_horse.rb:
```diff
@@ -0,0 +1,15 @@
+  class EvaluateHorse
+    def initialize(horse_races)
+      @horse_races = horse_races
+      @horse_evaluation = HorseEvaluation.new
```
意図しないサイド・エフェクトを避けるために、こちらの行は`call`で実行した方がいいです。

mterada1228 replied:
`call` メソッド内に処理を移動しました。

ershn replied:
OK

ershn commented on app/services/horse_stats/results_by_course_condition.rb:
```diff
@@ -0,0 +1,20 @@
+                                  horse_race.race.course_condition == course_condition
+                                end
+                              ))
+      end
```
ここで汎用的に書くと、読みにくくなって、保守性が下がるので、修正時に作業が増えるかもしれませんが、ここは明示的に書いた方がいいと思います。

```ruby
by_course_condition = horse_races.group_by { |horse_race| horse_race.race.course_condition }
@firm = HorsePodiums.new(by_course_condition[:firm])
# ...
```

mterada1228 replied:
https://github.com/mterada1228/keiba_expectation_app/blob/560e300c7b3444ab08cf86e3020068502d04d152/app/services/horse_stats/results_by_course_condition.rb#L9-L15 のように明示的に変数名を示すよう、修正しました。

ershn replied:
OK

ershn commented on app/services/horse_stats/horse_podiums.rb:
```diff
@@ -0,0 +1,21 @@
+      @second_place = horse_races_by_place[2].try(:count) || 0
+      @third_place = horse_races_by_place[3].try(:count) || 0
+      @unplaced = unplaced_horse_races(horse_races_by_place).try(:count) || 0
+    end
```
ここで５行だけなので、問題ないですが、それより多いと、全部を`initialize`の中に書くのは保守性が悪いので、以下にする方がいいです。

```ruby
def initialize(horse_races)
  @horse_races = horse_races
end

def horse_races_by_place
  @horse_races_by_place ||= horse_races.group_by(&:order_of_arrival)
end

def first_place
  @first_place ||= horse_races_by_place[1].try(:count) || 0
end

# ...
```

サイド・エフェクトのあるクラスで以上の書き方は逆に読みにくいですが、データを提供するだけの read only なモデルの場合、読みやすくなります。

mterada1228 replied:
https://github.com/mterada1228/keiba_expectation_app/blob/560e300c7b3444ab08cf86e3020068502d04d152/app/services/horse_stats/horse_podiums.rb#L7-L21 のようにメソッドとして別だししました。

ershn replied:
OK

ershn reviewed (changes requested):
ご確認ください。

mterada1228 committed:
refs features/show_stats_of_horse 指摘対応

ershn commented on app/controllers/horses_controller.rb:
```diff
@@ -7,8 +7,9 @@ def index
   def show
-    @horse = Horse.find(params[:id])
+    @horse = Horse.includes(:horse_races, :races).find(params[:id])
     @horse_races = @horse.horse_races.includes(:race).order('races.start DESC')
```
レコードの取得は`Horse.includes(horse_races: [:race])`で行えば、下の`includes(:race)`を避けられるかもしれません。

order 区は該当のレコードはすでに取得されていれば、ruby でやっても問題なさそうです。

mterada1228 replied:
頂きましたご指摘の通り修正しました。
実際に request しましたが、SQL の呼び出しは一度だけしか実行されないこと確認できました。

ershn replied:
OK

ershn commented on app/services/horse_stats/evaluate_horse.rb:
```diff
@@ -0,0 +1,15 @@
+      @horse_evaluation = HorseEvaluation.new
+      @horse_evaluation.results_by_course_condition =
+        HorseStats::ResultsByCourseCondition.new(horse_races: @horse_races)
+      @horse_evaluation
```
`@horse_evaluation`はインスタンス変数にしなくても問題ないと思います。

mterada1228 replied:
https://github.com/mterada1228/keiba_expectation_app/pull/16#discussion_r640286986

上記コメントに対する修正により、 `@horse_evaluation` への代入は不必要になりましたので削除しています。

ershn replied:
OK

ershn commented on app/services/horse_stats/results_by_course_condition.rb:
```diff
@@ -0,0 +1,17 @@
+      create_stats(horse_races)
+    end
+
+    def create_stats(horse_races)
```
こちらは`private`にしても問題なさそうです。

mterada1228 replied:
`private` に変更しました。

ershn replied:
OK

ershn commented on app/services/horse_stats/horse_podiums.rb:
```diff
@@ -0,0 +1,35 @@
+    private
+
+    def by_place
+      @by_place ||= @horse_races ? @horse_races.group_by(&:order_of_arrival) : {}
```
クラスがパブリックインターフェイスで受け取れるデータの形とパターンは絞った方が保守性がいいので、このクラスで`nil`を受け取れないと決めて、呼び出し箇所で必ず配列を渡すようにしたほうがいいです。

```ruby
@firm = HorsePodiums.new(by_course_condition.fetch('firm', []))
@good = HorsePodiums.new(by_course_condition.fetch('good', []))
```

mterada1228 replied:
頂いたご指摘の通り修正させて頂きました。

ershn replied:
OK

ershn commented on app/services/horse_stats/horse_podiums.rb:
```diff
@@ -0,0 +1,35 @@
+module HorseStats
+  class HorsePodiums
+    def initialize(horse_races)
```
引数として配列しか許容しないと他メンバーに伝えたい場合はコメントでそれを説明するのが一般的です。

フォーマットとしては最近の流れからして YARD がいいと思います。

mterada1228 replied:
YARD でコメントを追加しました。

ershn replied:
OK

ershn commented on app/services/horse_stats/horse_evaluation.rb:
```diff
@@ -0,0 +1,6 @@
+# 馬別統計データクラス
+module HorseStats
+  class HorseEvaluation
+    attr_accessor :results_by_course_condition
```
ここでレコードのようにいつどこでクラスのステートを変えてもいいような実装ですが、それはできる場合に避けて、使い手に選択肢が少ない実装にした方が保守性が上がります。

ここはすべてをクラスのインスタンス化時にデータを渡せばいいと思います。

```ruby
def initialize(results_by_course_condition:)
end

attr_reader :results_by_course_condition
```

mterada1228 replied:
`initialize` メソッドを追加しました。
それに伴い、呼び出し側も以下のように修正しました。

https://github.com/mterada1228/keiba_expectation_app/blob/8d0f7758e696bc6f4cd0dd8ead1b9d9db68a33f8/app/services/horse_stats/evaluate_horse.rb#L8-L13

ershn replied:
OK

mterada1228 committed:
refs features/show_stats_of_horse 指摘対応

ershn commented on app/services/horse_stats/horse_evaluation.rb:
```diff
@@ -0,0 +1,10 @@
+    attr_reader :results_by_course_condition
+
+    def initialize(results_by_course_condition:)
+      @results_by_course_condition = results_by_course_condition
```
`results`はデータの形について何も指定しない呼称なので、`podiums`と名付けた方がわかりやすいと思いました。

mterada1228 replied:
名称を変更しました。

ershn commented on app/controllers/horses_controller.rb:
```diff
@@ -7,8 +7,9 @@ def index
-    @horse = Horse.find(params[:id])
-    @horse_races = @horse.horse_races.includes(:race).order('races.start DESC')
+    @horse = Horse.includes(horse_races: [:race]).find(params[:id])
+    @horse_races = @horse.horse_races.sort_by { |horse_race| horse_race.race.start }.reverse!
```
`!`があるメソッドだと、リシーバーの配列が変更されますが、パフォーマンスが問題じゃなければ基本的に`!`のないメソッドを使ったほうが保守性がよくなります。

----

ここは`sort_by`+`reverse`より`sort { |a, b| }`を使ったほうがパフォーマンスがいいと思います。

mterada1228 replied:
`sort { |a, b|  b.race.start <=> a.race.start }` でレース日付の降順に並ぶようにロジックを修正しました。

ershn reviewed (changes requested):
ご確認ください。

mterada1228 committed:
指摘対応
  - result_by_course_condition は podiums_by_course_condition に修正
  - HorseRace をレース日付の降順に並べるロジックは、sort_by + reverseではなく、sort に変更
